src/devtools/pachdev/actions.go

```diff
@@ -0,0 +1,571 @@
+
+	// kindProvider is a pre-initialized kind client (or cluster provider,
+	// actually, as kind is local), to be shared across various subtasks. Like
+	// ctx, pachClient, and others, this is initialized lazily by the
+	// KindProvider() method, and should only be accessed through it (even
+	// internally, by other methods, in case it hasn't been initialized yet).
+	// This is so that it's not initialized for operations that don't require
+	// it, such as printing out a manifest, allowing those operations to be run
+	// even in the absence of kind or Docker.)
+	kindProvider *cluster.Provider
+
+	// k8sManifest is the kubernetes manifest that will be deployed. Like ctx,
+	// pachClient, and others, This is initialized lazily by K8sManifest() and
+	// should only be accessed through it (even internally, by other methods, in
+	// case it hasn't been initialized yet).
+	k8sManifest string
+
+	// err is set if the overall operation has failed, and indicates the cause.
+	err error
+
+	// oldPachdPodName is the name of the pachd pod that will be replaced
+	// (including pod ID and so on). This is used to determined whether 'pachctl
+	// version' is talking to the old pod or the new pod.
+	oldPachdPodName string
+
+	// oldPachVersion is the version of pachyderm that is being replaced. If
+	// this is the same as 'pachVersion', then we don't need to push a new
+	// image.
+	oldPachVersion string
+}
+
+// Option is the type of optional values passed to NewDeployOp
+type Option func(*deployOp) error
+
+// WithPachVersion is an option for NewDeployOp that sets the version of the
+// Pachyderm instance that will be deployed. The argument can be any released
+// version of Pachyderm (e.g. "2.8.1"), or any of the special strings:
+//   - "local" (indicating that a locally-build Pachyderm image should be
+//     deployed; commonly used on Core)
+//   - "latest" (indicating that the most recent minor release of Pachyderm should
+//     be deployed) (TODO)
+//   - nightly (indicating that the most recent nightly release of Pachyderm
+//     should be deployed)
+func WithPachVersion(v string) Option {
+	return func(op *deployOp) error {
+		op.pachVersion = v
+		return nil
+	}
+}
+
+func WithObjectStorage(v string) Option {
+	// WithPachVersion is an option for NewDeployOp that sets the version of the
+	// Pachyderm instance that will be deployed.
+	return func(op *deployOp) error {
+		op.pachVersion = v
+		return nil
+	}
+}
+
+func NewDeployOp(verbose bool, opts ...Option) (*deployOp, error) {
+	result := &deployOp{
+		pachVersion:   "local",
+		objectStorage: "minio",
+	}
+	for _, op := range opts {
+		if err := op(result); err != nil {
+			return nil, err
+		}
+	}
+
+	return result, nil
+}
+
+// ctx and cancel are captured by result.Ctx(), and will not be initialized
+// if that is never called
+func (op *deployOp) Ctx() context.Context {
+	if op.ctx != nil {
+		return op.ctx
+	}
+	op.ctx, op.cancel = context.WithCancel(context.Background())
```

Comment:
Consider pctx.WithCancel.  The cancellation error message changes from "context cancelled" to "context cancelled at foo.go:42, created at bar.go:123".